### PR TITLE
Merge Pycorels changes

### PR DESCRIPTION
**Pull request by fingoldin**
Please let me know if you think these changes are acceptable! I'd be happy to explain the reasoning for them. Among other reasons, I made these changes to:
- Compile and run without GMP
- Compile and run on Windows
- Change verbosity options
- Clean up memory leaks
- Improve debug output

### CODE REVIEW

File: src/GNUmakefile [hunk 1]
```diff
@@ -1,66 +1,43 @@
...
-CXXFLAGS := $(CFLAGS) -std=c++11
-
-ifneq (,$(filter $(shell uname -n),elaines-MacBook-Pro.local Nicholass-Air-5.wireless.yale.internal))
-    LIBS := -L/usr/local/lib
-    CFLAGS += -I/usr/local/include
-else
-    LIBS :=
+CFLAGS := -g -W -Wall -Wno-unused-parameter -I. -O2 -fno-omit-frame-pointer
+LIBS := 
+ifndef NGMP
```
Comment (nlarusstone) [inline src/GNUmakefile:3]: what's NGMP?

File: src/GNUmakefile [hunk 2]
```diff
@@ -1,66 +1,43 @@
...
+    CFLAGS += -I/usr/local/include -DGMP
```
Comment (nlarusstone) [inline src/GNUmakefile:4]: Does this work on Linux as well as Mac?

File: src/GNUmakefile [hunk 3]
```diff
@@ -1,66 +1,43 @@
...
-    CFLAGS += -fno-omit-frame-pointer
-else
-    LDFLAGS :=
-endif
-
-ifneq ($(shell uname),Darwin)
-    LIBS += #-lbsd
-endif
-
-CC = gcc
```
Comment (nlarusstone) [inline src/GNUmakefile:24]: I think we still want to use gcc instead of g++ to compile rulelib.c

File: src/cache.cpp [hunk 4]
```diff
@@ -43,7 +47,7 @@ CacheTree::CacheTree(size_t nsamples, size_t nrules, double c, rule_t *rules,
 }
 
 CacheTree::~CacheTree() {
-    if (root_)
```
Comment (nlarusstone) [inline src/cache.cpp:46]: We originally had this in here because it was segfaulting, so I'd want to make sure we test this behavior carefully

File: src/cache.cpp [hunk 5]
```diff
@@ -69,11 +73,11 @@ Node* CacheTree::construct_node(unsigned short new_rule, size_t nrules, bool pre
  * Inserts the root of the tree, setting up the default rules.
  */
 void CacheTree::insert_root() {
-    VECTOR tmp_vec;
+    //VECTOR tmp_vec;
     size_t d0, d1;
     bool default_prediction;
     double objective;
-    make_default(&tmp_vec, nsamples_);
+    //make_default(&tmp_vec, nsamples_);
```
Comment (nlarusstone) [inline src/cache.cpp:80]: If nsamples_ isn't initialized, how can you use it d1?

File: src/corels.cpp [hunk 6]
```diff
@@ -145,25 +152,23 @@ void evaluate_children(CacheTree* tree, Node* parent, tracking_vector<unsigned s
     }
 }
 
+static size_t num_iter = 0;
+static double min_objective = 0.0;
+static VECTOR captured, not_captured;
```
Comment (nlarusstone) [inline src/corels.cpp:157]: I'd prefer for these to not be global variables -- any chance we can move them back into the function?

File: src/corels.cpp [hunk 7]
```diff
@@ -175,79 +180,89 @@ int bbound(CacheTree* tree, size_t max_num_nodes, Queue* q, PermutationMap* p) {
...
-        }
-        logger->setQueueSize(q->size());
-        if (queue_min_length < logger->getQueueMinLen()) {
-            // garbage collect the permutation map: can be simplified for the case of BFS
-            queue_min_length = logger->getQueueMinLen();
-            //pmap_garbage_collect(p, queue_min_length);
-        }
-        ++num_iter;
-        if ((num_iter % 10000) == 0) {
+void bbound_loop(CacheTree* tree, Queue* q, PermutationMap* p) {
```
Comment (nlarusstone) [inline src/corels.cpp:185]: I love that you separated this into a function -- the merging with our parallel code might be a bit tricky, but nice functional decomposition!

File: src/main.cc [hunk 8]
```diff
@@ -3,21 +3,20 @@
...
-NullLogger* logger;
+#ifndef _WIN32
+#define _snprintf snprintf
+#endif
 
 int main(int argc, char *argv[]) {
     const char usage[] = "USAGE: %s [-b] "
-        "[-n max_num_nodes] [-r regularization] [-v verbosity] "
-        "-c (1|2|3|4) -p (0|1|2) [-f logging_frequency]"
+        "[-n max_num_nodes] [-r regularization] [-v (rule|label|minor|samples|progress|loud|silent)] "
```
Comment (nlarusstone) [inline src/main.cc]: If verbosity is using these enums, then why are we checking ints in the code above?

File: src/main.cc [hunk 9]
```diff
@@ -59,7 +63,16 @@ int main(int argc, char *argv[]) {
             use_captured_sym_map = map_type == 2;
             break;
         case 'v':
-            verbosity = atoi(optarg);
+            verb_trim = strtok(optarg, " ");
```
Comment (nlarusstone) [inline src/main.cc]: This code is a bit hard to parse -- any chance we could factor out into a helper function and then test that function?

File: src/main.cc [hunk 10]
```diff
@@ -59,7 +63,16 @@ int main(int argc, char *argv[]) {
...
+            strcpy(verbstr, verb_trim);
+            vopt = strtok(verb_trim, ",");
+            while (vopt != NULL) {
+                if (!strstr(vstr, vopt)) {
+                    verr = true;
```
Comment (nlarusstone) [inline src/main.cc]: should error be true here as well?

File: src/main.cc [hunk 11]
```diff
@@ -122,110 +161,95 @@ int main(int argc, char *argv[]) {
...
+    }
+
+    if(nlabels != 2) {
+        fprintf(stderr, "nlabels must be equal to 2, got %d\n", nlabels);
+        rules_free(rules, nrules, 1);
+        rules_free(labels, nlabels, 0);
+        return 1;
+    }
+
+    if(nsamples_chk != nsamples) {
```
Comment (nlarusstone) [inline src/main.cc]: Nice!

File: src/main.cc [hunk 12]
```diff
@@ -122,110 +161,95 @@ int main(int argc, char *argv[]) {
...
+        fprintf(stderr, "nsamples mismatch between out file (%d) and label file (%d)\n", nsamples, nsamples_chk);
+        rules_free(rules, nrules, 1);
+        rules_free(labels, nlabels, 0);
+        return 1;
+    }
 
     int nmeta, nsamples_check;
```
Comment (nlarusstone) [inline src/main.cc]: Can we rename nsamples_check to something relating to meta

File: src/main.cc [hunk 13]
```diff
@@ -122,110 +161,95 @@ int main(int argc, char *argv[]) {
...
-        CapturedPermutationMap* cap_pmap = new CapturedPermutationMap;
-        p = (PermutationMap*) cap_pmap;
+        int* rulelist = NULL;
+        int rulelist_size = 0;
+        int* classes = NULL;
+
+        run_corels_end(&rulelist, &rulelist_size, &classes, 0, latex_out, rules, labels, &opt_fname[0]);
+
+        if(rulelist)
+            free(rulelist);
```
Comment (nlarusstone) [inline src/main.cc]: Are you mallocing these? It's usually good form to put the allocation near the definition so it's clear that they're on the heap.

File: src/rulelib.cpp [hunk 14]
```diff
@@ -90,23 +87,23 @@ rules_init(const char *infile, int *nrules,
 	 * the end.
 	 */
 	rule_cnt = add_default_rule != 0 ? 1 : 0;
-	while (getline(&line, &len, fi) != -1) {
+	while (m_getline(&line, &len, fi) != -1) {
```
Comment (nlarusstone) [inline src/rulelib.cpp]: What's m_getline and does it have different behavior?

Comment (nlarusstone) [reply src/rulelib.cpp]: Same with all the m_ functions

File: src/rulelib.cpp [hunk 15]
```diff
@@ -281,7 +284,7 @@ ascii_to_vector(char *line, size_t len, int *nsamples, int *nones, VECTOR *ret)
 	ones = 0;
 
 
-	for(p = line; len-- > 0; p++) {
+	for(p = line; len-- > 0 && *p != '\0'; p++) {
```
Comment (nlarusstone) [inline src/rulelib.cpp:291]: What if len is > 0 but *p == '\0'? Then we'd have a bug and probably want to throw an error, no?

File: src/rulelib.cpp [hunk 16]
```diff
@@ -306,20 +309,22 @@ ascii_to_vector(char *line, size_t len, int *nsamples, int *nones, VECTOR *ret)
 	}
 
 	/* Store val if it contains any bits. */
+    // Changed to make the last non-full v_entry left-aligned
```
Comment (nlarusstone) [inline src/rulelib.cpp:316]: Why?

File: src/rulelib.cpp [hunk 17]
```diff
@@ -595,59 +601,6 @@ ruleset_delete(rule_t *rules, int nrules, ruleset_t *rs, int ndx)
 	return;
 }
 
-/*
- * We create random rulesets for testing and for creating initial proposals
- * in MCMC
- */
-int
-create_random_ruleset(int size,
```
Comment (nlarusstone) [inline src/rulelib.cpp:603]: Why did you remove this?

File: src/rulelib.cpp [hunk 18]
```diff
@@ -905,7 +858,7 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
 	int total_support;
 
 	printf("%d rules %d samples\n", rs->n_rules, rs->n_samples);
-	n = (rs->n_samples + BITS_PER_ENTRY - 1) / BITS_PER_ENTRY;
```
Comment (nlarusstone) [inline src/rulelib.cpp:908]: Is this incorrect?

File: src/rulelib.cpp [hunk 19]
```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
...
 {
 	rule_t *r;
 
 	r = rules + ndx;
-	printf("RULE %d: ( %s ), support=%d, card=%d:",
+	printf("RULE %d: ( %s ), support=%d, card=%d",
 	    ndx, r->features, r->support, r->cardinality);
-	if (detail)
-		rule_vector_print(r->truthtable, n);
+	if (detail) {
```
Comment (nlarusstone) [inline src/rulelib.cpp]: Indentation looks strange here

File: src/rulelib.cpp [hunk 20]
```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
...
+	char* str = mpz_get_str(NULL, 2, v);
+    int len = strlen(str);
+    for(int i = 0; i < (nsamples - len); i++) {
+        fputc('0', stdout);
+    }
+    fputs(str, stdout);
+	fputc('\n', stdout);
 #else
-	for (int i = 0; i < n; i++)
+	/*for (int i = 0; i < n; i++)
```
Comment (nlarusstone) [inline src/rulelib.cpp]: Remove commented code if we're not using

File: src/rulelib.cpp [hunk 21]
```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
...
+*/
+	return !!(v[e / BITS_PER_ENTRY] & (one << shift));
+#endif
+}
+
+void
+rule_set(VECTOR v, int e, int val, int n) {
+#ifdef GMP
+    if(val)
+	    mpz_setbit(v, e);
```
Comment (nlarusstone) [inline src/rulelib.cpp]: indentation

File: src/rulelib.cpp [hunk 22]
```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
...
+	    v[e/BITS_PER_ENTRY] &= ((v_entry) -1) - (one << shift);
 #endif
 }
 
 int
-rule_vector_equal(const VECTOR v1, const VECTOR v2, short len1, short len2) {
+rule_vector_cmp(const VECTOR v1, const VECTOR v2, int len1, int len2) {
 #ifdef GMP
-    return !mpz_cmp(v1, v2);
+    return mpz_cmp(v1, v2);
```
Comment (nlarusstone) [inline src/rulelib.cpp:988]: Whoaa this is a big change -- did you make sure to find everywhere in the code we used it?

Comment (nlarusstone) [reply src/rulelib.cpp:988]: Also, why the change?

File: src/rulelib.cpp [hunk 23]
```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
...
 #else
     if (len1 != len2)
-        return 0;
+        return 2 * (len1 > len2) - 1;
```
Comment (nlarusstone) [inline src/rulelib.cpp:991]: Return 1 or -1? Why not 1 or 0?

File: src/run.hh [hunk 24]
```diff
@@ -0,0 +1,22 @@
+#ifndef RUN_H
+#define RUN_H
+
+#include "rule.hh"
+
+#ifdef __cplusplus
```
Comment (nlarusstone) [inline src/run.hh]: what if it's not defined?

File: src/utils.cpp [hunk 25]
```diff
@@ -91,15 +104,97 @@ std::string Logger::dumpPrefixLens() {
...
-            s += std::to_string(_state.prefix_lens[i]);
+            s += sizet_tostring(_state.prefix_lens[i]);
             s += ";";
         }
     }
     return s;
 }
 
+// Custom getline for Windows support
+signed long long m_getline(char** lineptr, size_t* n, FILE* stream)
```
Comment (nlarusstone) [inline src/utils.cpp]: Oof do we need custom functions or can we use a Microsoft implementation (a la https://social.msdn.microsoft.com/Forums/vstudio/en-US/94bc15ac-1c91-4252-977e-ea010c3e9d5f/the-usage-of-getline?forum=vclanguage)

File: src/run.cpp [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+#include <iostream>
+#include <set>
+
+#include "queue.hh"
+#include "run.hh"
+
+#define BUFSZ 512
+
+NullLogger* logger = nullptr;
+static PermutationMap* g_pmap = nullptr;
```
Comment (nlarusstone) [inline src/run.cpp]: Can we avoid globals here?

File: src/run.cpp [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+static CacheTree* g_tree = nullptr;
+static Queue* g_queue = nullptr;
+static double g_init = 0.0;
+static std::set<std::string> g_verbosity;
+
+int run_corels_begin(double c, char* vstring, int curiosity_policy,
+                  int map_type, int ablation, int calculate_size, int nrules, int nlabels,
+                  int nsamples, rule_t* rules, rule_t* labels, rule_t* meta, int freq, char* log_fname)
+{
+    g_verbosity.clear();
```
Comment (nlarusstone) [inline src/run.cpp]: What's the g_ prefix?

File: src/run.cpp [hunk 28]
```diff
@@ -0,0 +1,192 @@
+#include <stdio.h>
```
Comment (nlarusstone) [inline src/run.cpp:1]: I might name this something other than run.cpp since it contains much of our core logic and are not just runner scripts

File: src/run.cpp [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+        printf("\n\n");
+    }
+
+    if (g_verbosity.count("minor") && meta) {
+        printf("Minority bound for %d samples\n\n", nsamples);
+        rule_print_all(meta, 1, nsamples, g_verbosity.count("samples"));
+        printf("\n\n");
+    }
+
+    if (g_tree)
```
Comment (nlarusstone) [inline src/run.cpp]: Get rid of these when you remove the globals

File: src/run.cpp [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+    const tracking_vector<unsigned short, DataStruct::Tree>& r_list = g_tree->opt_rulelist();
+    const tracking_vector<bool, DataStruct::Tree>& preds = g_tree->opt_predictions();
+
+    double accuracy = 1.0 - g_tree->min_objective() + g_tree->c() * r_list.size();
+
+    *rulelist = (int*)malloc(sizeof(int) * r_list.size());
+    *classes = (int*)malloc(sizeof(int) * (1 + r_list.size()));
+    *rulelist_size = r_list.size();
+    for(size_t i = 0; i < r_list.size(); i++) {
+        (*rulelist)[i] = r_list[i];
```
Comment (nlarusstone) [inline src/run.cpp]: rulelist->i should be equivalent

**Review by nlarusstone — CHANGES_REQUESTED** [review]

File: src/GNUmakefile (hunk 1 above)
Comment (fingoldin) [reply src/GNUmakefile:3]: It's a flag to compile without GMP. You can run 'make NGMP=1' to compile everything without GMP.

File: src/GNUmakefile (hunk 2 above)
Comment (fingoldin) [reply src/GNUmakefile:4]: Yes, for both Python 2.7 and Python 3.7

File: src/cache.cpp (hunk 4 above)
Comment (fingoldin) [reply src/cache.cpp:46]: Yeah, I tested it very carefully. As you had it, it worked fine with GMP, but without GMP for some reason it would segfault when the optimal rulelist was just the default prediction. Replacing '_root' with 'num_nodes()' fixed that, and now it doesn't segfault with GMP or without GMP.

File: src/cache.cpp (hunk 5 above)
Comment (fingoldin) [reply src/cache.cpp:80]: It's initialized in line 31 of cache.cpp, in CacheTree's constructor

File: src/corels.cpp (hunk 7 above)
Comment (fingoldin) [reply src/corels.cpp:185]: Thank you :D

File: src/corels.cpp (hunk 6 above)
Comment (fingoldin) [reply src/corels.cpp:157]: I dislike how it is right now as well, but this was the first way I could think of to have a bbound_loop function that depends on the same num_iter, min_objective, etc. variables as bbound_begin and bbound_end. I suppose we could pass them as pointer arguments to those three functions instead, what do you think? Having them as static ensures they never leave corels.cpp, though.

File: src/main.cc (hunk 8 above)
Comment (fingoldin) [reply src/main.cc]: Sorry, I'm not sure I understand - what code above are you referring to? There is also a verbosity integer passed to the Logger, which specifies whether a log file should be generated, and how much progress should be printed. The value of the integer depends on the enums (see line 79 of run.cpp).

File: src/main.cc (hunk 9 above)
Comment (fingoldin) [reply src/main.cc]: Sure, I will put it into a separate function. By testing it do you mean making unit tests for the command line only program?

File: src/main.cc (hunk 10 above)
Comment (fingoldin) [reply src/main.cc]: error gets set to true if verr is true in line 122.

File: src/GNUmakefile (hunk 3 above)
Comment (fingoldin) [reply src/GNUmakefile:24]: This would require creating a new pair of files to house my implementations of getline, strsep, and strdup (since they are currently used in rulelib.cpp, but declared in a c++ header file utils.hh). There might still be problems though, since I remember I moved rulelib.c to rulelib.cpp to make the build succeed on Windows.

File: src/main.cc (hunk 10 above)
Comment (fingoldin) [reply src/main.cc]: Now line 131

File: src/main.cc (hunk 11 above)
Comment (fingoldin) [reply src/main.cc]: Yee

File: src/main.cc (hunk 12 above)
Comment (fingoldin) [reply src/main.cc]: Done (for nsamples_chk too)

File: src/main.cc (hunk 13 above)
Comment (fingoldin) [reply src/main.cc]: I malloc them in run_corels_end because I can't know their size before running bbound_end within run_corels_end. I could malloc them in main.cc with as a fixed-width buffer, but then there'd be an upper bound on the length of the rulelist that could be outputted.

File: src/rulelib.cpp (hunk 14 above)
Comment (fingoldin) [reply src/rulelib.cpp]: So those were three functions that weren't present on Windows, as I think they're posix functions. In order to compile on Windows, I had to write my own implementations of them; see line 117 of utils.cpp.

File: src/rulelib.cpp (hunk 15 above)
Comment (fingoldin) [reply src/rulelib.cpp:291]: Yeah, for some reason that bug was present in some data files after compiling without GMP so I added that condition. I spend a bunch of time but simply couldn't find the bug. Fortunately, like this, if the data file actually has a problem it will be handled by line 318. Specifically, if one line is too short or empty, the loop will terminate early, and 'i' will be less than '*nsamples', triggering an error and causing the program to fail.

File: src/rulelib.cpp (hunk 16 above)
Comment (fingoldin) [reply src/rulelib.cpp:316]: Because every v_entry would be filled left-aligned, but in such a way that if the last v_entry was partially empty it would be right aligned (because the left-shift only happens as long as there are bits to assign). This ensures that rule_isset and rule_set work properly without GMP.

File: src/rulelib.cpp (hunk 17 above)
Comment (fingoldin) [reply src/rulelib.cpp:603]: Because the RANDOM_RANGE macro would cause compilation errors on Windows, and we weren't using this function anywhere.

File: src/rulelib.cpp (hunk 18 above)
Comment (fingoldin) [reply src/rulelib.cpp:908]: In order to properly print the right number of samples in rule_vector_print, I needed to know the exact amount of samples. Previously, rule_vector_print was only being passed the number of v_entries in non-GMP mode, which would cause it to print 128 samples when there were actually anywhere between 65 and 128. Similarly, if the bitvector had zeroes in the most significant bits, those weren't printed in GMP mode, making the bitvector look shorter than it actually was. So, I had to propagate nsamples through all of the printing functions instead of the number of v_entries.

File: src/rulelib.cpp (hunk 19 above)
Comment (fingoldin) [reply src/rulelib.cpp]: Thanks for pointing that out, I must have edited it on Windows or something.

File: src/rulelib.cpp (hunk 20 above)
Comment (fingoldin) [reply src/rulelib.cpp]: Done

File: src/rulelib.cpp (hunk 21 above)
Comment (fingoldin) [reply src/rulelib.cpp]: Thanks again

File: src/rulelib.cpp (hunk 22 above)
Comment (fingoldin) [reply src/rulelib.cpp:988]: The only places rule_vector_equal (which I renamed to rule_vector_cmp) was used were lines 61 and 63 of pmap.hh. I made sure to change those. The reason for the change was because I needed a comparison function that returned 0, -1, or 1 depending on which of two vectors was greater (I used it in the minority bound generation, see https://github.com/fingoldin/pycorels/blob/master/corels/src/utils.cpp), and since rule_vector_equal was essentially what I needed, I renamed it to rule_vector_cmp and made it return 0, -1, or 1.

File: src/rulelib.cpp (hunk 23 above)
Comment (fingoldin) [reply src/rulelib.cpp:991]: See comment in previous conversation.

File: src/run.hh (hunk 24 above)
Comment (fingoldin) [reply src/run.hh]: This was left over from when I was compiling main.cc as C. I'll remove it.

File: src/utils.cpp (hunk 25 above)
Comment (fingoldin) [reply src/utils.cpp]: strsep and strdup were simple enough that I could write my own versions quickly. Using Microsoft's implementation of getline would work, but then we'd have to use C++ code in rulelib.cpp, which is all C at the moment (and I thought that'd be kinda ugly). If you think it's better, I can go ahead and do it. My implementation of getline seems to be working fine, however, on Windows, Mac, and Linux.

File: src/run.cpp (hunk 27 above)
Comment (fingoldin) [reply src/run.cpp]: 'Global'

File: src/run.cpp (hunk 30 above)
Comment (fingoldin) [reply src/run.cpp]: Hmm, that doesn't seem to work for me - it gives an error:
error: request for member ‘i’ in ‘* rulelist’, which is of non-class type ‘int*’

File: src/run.cpp (hunk 28 above)
Comment (fingoldin) [reply src/run.cpp:1]: Since corels is taken and the main() function is in main.cc, how about "algo.cpp"? I named it run.cpp because it houses the three run functions, run_corels_begin, run_corels_loop, run_corels_end.

File: src/run.cpp (hunk 26 above)
Comment (fingoldin) [reply src/run.cpp]: I removed the globals, with the magic of C++ references!

File: src/run.cpp (hunk 29 above)
Comment (fingoldin) [reply src/run.cpp]: Isn't it easier to have the free code here? Otherwise, I'd have to copy it to both main.cc and _corels.pyx (the Cython file that calls run_corels_end from Python).

File: src/run.cpp (hunk 27 above)
Comment (fingoldin) [reply src/run.cpp]: I removed the 'g_' prefixes

File: src/corels.cpp (hunk 6 above)
Comment (fingoldin) [reply src/corels.cpp:157]: UPDATE: I removed the global variables from run.cpp by making them arguments. I could make these arguments as well, but then the three run functions would have a ridiculously long string of arguments (since these would have to be passed to the run functions as well as the bbound functions). What do you think?

File: src/cache.cpp (hunk 5 above)
Comment (nlarusstone) [reply src/cache.cpp:80]: Let's remove the commented code then.

File: src/corels.cpp (hunk 6 above)
Comment (nlarusstone) [reply src/corels.cpp:157]: I think it's ok for now, but we should think about how we can incorporate them as either arguments or attributes of an object.

File: src/main.cc (hunk 13 above)
Comment (nlarusstone) [reply src/main.cc]: Why don't you make them vectors so you can add to them dynamically? This would align with the types we use for `get_prefix_and_predictions` more closely anyways

File: src/run.cpp (hunk 30 above)
Comment (nlarusstone) [reply src/run.cpp]: This will change if we make this a vector (which I think we should)

File: src/run.cpp (hunk 28 above)
Comment (nlarusstone) [reply src/run.cpp:1]: Ok, we can keep it as is for now

File: src/utils.cpp (hunk 25 above)
Comment (nlarusstone) [reply src/utils.cpp]: I would really really prefer to use library functions if possible. Since you moved rulelib.cpp to be compiled by g++ anyways, I think we should go ahead and use the Microsoft implementations for all of these functions. We can wrap them in `#ifdef`s to tell if we're on Windows, no?

File: src/rulelib.cpp (hunk 14 above)
Comment (nlarusstone) [reply src/rulelib.cpp]: See my comments in utils.cpp

File: src/utils.cpp [hunk 31]
```diff
@@ -70,19 +79,23 @@ void Logger::dumpState() {
...
+#ifdef GMP
 /*
  * Uses GMP library to dump a string version of the remaining state space size.
  * This number is typically very large (e.g. 10^20) which is why we use GMP instead of a long.
  * Note: this function may not work on some Linux machines.
  */
 std::string Logger::dumpRemainingSpaceSize() {
-    mpz_class s(_state.remaining_space_size);
-    return s.get_str();
+    char* str = mpz_get_str(NULL, 10, _state.remaining_space_size);
```
Comment (nlarusstone) [inline src/utils.cpp:93]: Can we wrap in a ifdef GMP?

File: src/rulelib.cpp [hunk 32]
```diff
@@ -71,9 +69,9 @@ int
 rules_init(const char *infile, int *nrules,
     int *nsamples, rule_t **rules_ret, int add_default_rule)
 {
-	FILE *fi;
-	char *line = NULL;
-    char *rulestr;
+	std::ifstream fi(infile);
+  std::string line;
```
Comment (nlarusstone) [inline src/rulelib.cpp]: Indentation looks a bit strange here

File: src/utils.cpp (hunk 31 above)
Comment (fingoldin) [reply src/utils.cpp:93]: Shouldn't this do the same thing? It's just getting the remaining state size as a string in base 10 format. With this we don't have to link with a whole new library (gmpxx)

File: src/rulelib.cpp (hunk 32 above)
Comment (fingoldin) [reply src/rulelib.cpp]: Did that fix it?

File: src/utils.cpp (hunk 31 above)
Comment (nlarusstone) [reply src/utils.cpp:93]: Ah sorry I missed the ifdef on line 86 -- looks good here!

File: src/rulelib.cpp (hunk 32 above)
Comment (nlarusstone) [reply src/rulelib.cpp]: Hmmm now it looks like some of them are indented with 2 spaces and some with 8...

Comment (fingoldin) [reply src/rulelib.cpp]: Ok I see what the problem was.

File: src/main.cc (hunk 9 above)
Comment (fingoldin) [reply src/main.cc]: Ok, I used a handy C++ unit test library named Catch (https://github.com/catchorg/Catch2) to make some unit tests for this function. I also set it up so that we can continue to add unit tests for the whole command line program in the future. To run the tests, do 'make tests' from the src folder, followed by './tests'.

**Review by nlarusstone — DISMISSED** [review]

**Review by nlarusstone — APPROVED** [review]
